File: ABC054/ABC054C.cpp
#include <iostream>
#include <iomanip>
#include <cmath>
#include <algorithm>
#include <functional>
#include <stack>
#include <queue>
#include <string>
#include <unordered_map>

using namespace std;
typedef long long ll;
typedef pair<ll,ll> P;

ll N,M,ans;
const ll nmax=8;
const ll mmax=8*(8-1)/2;
ll graph[nmax+1][nmax+1];
bool visited[nmax],allvisited;

void dfs(ll c){
    allvisited=true;
//    cout<<N<<" "<<M<<endl;
    for(ll i=1;i<=N;i++){
        if(!visited[i]){
            allvisited=false;
            break;
        }
    }
    if(allvisited){
//        cout<<"aaa"<<endl;
        ans++;
        //return;
    }
    for(ll i=1;i<=N;i++){
        if(graph[c][i]==1&&(!visited[i])){
            visited[i]=true;
            dfs(i);
            visited[i]=false;
        }
    }
}

int main(){
    cin>>N>>M;
    ll a,b;
    for(int i=0;i<M;i++){
        cin>>a>>b;
        graph[a][b]=graph[b][a]=1;
    }
    visited[1]=true;
    dfs(1);
    cout<<ans<<endl;
    return 0;
}

File: ABC054/ABC054C.py
def dfs(v,N,visited):
    allvis=True
    for i in range(N):
        if visited[i]==False:
            allvis=False
    if allvis:
        return 1
    ret=0

    for i in range(N):
        if graph[v][i]==False:
            continue
        if visited[i]:
            continue
        visited[i]=True
        ret+=dfs(i,N,visited)
        visited[i]=False
    return ret



N,M = map(int, input().strip().split())
graph = [[False] * N for i in range(N)]
for i in range(M):
    a,b=map(int,input().strip().split())
    graph[a-1][b-1]=True
    graph[b-1][a-1]=True


visited=[False]*N
visited[0]=True

print(dfs(0,N,visited))
